### PR TITLE
Add task to convert docx files to pdf's

### PR DESCRIPTION
**Pull request by jutonz**
This connects code from a few different areas to allow processing all SourceFiles with docx attachments and converting them into pdf's. The docx file is preserved as `source_file.active_storage_attachment`--i.e. it's not changed at all--and the new PDF is uploaded as `source_file.redacted_file`.

I didn't attempt to convert SourceFiles which
* already have a redacted file attachment, or
* have a docx file with embedded attachments

Right now this is also just a script that you can run once. It's idempotent so you should be able to run it over and over if need be, but because we need to download the file to check if it has attachments it will still try to reprocess those files each time you run the script. I thought about adding a `pdf_conversion_attempted_at` column or something but opted not to for now. lmk if you feel differently though!

I think maybe a next step would be to run this in the background each time a source file is created so that all happens automatically, and then we could even just delete this rake task eventually.

[Asana ticket](https://app.asana.com/0/1203289004376659/1206305102472876)

### CODE REVIEW

File: app/models/source_file.rb [hunk 1]
```diff
@@ -10,6 +10,45 @@ class SourceFile < ApplicationRecord
...
+    includes(active_storage_attachment: :blob).where(
+      active_storage_attachment: {
+        active_storage_blobs: {
+          content_type: PDF_CONTENT_TYPE
+        }
+      }
+    )
+  end
+
+  def self.docx_attachment
```
[the remark this answers is not in the document]
Comment (FerPerales) [reply app/models/source_file.rb:23]: @jutonz thanks for pointing this! Do you know if there's a way we can enforce it using stadardrb? Would be nice to catch these automatically

File: app/services/doc_to_pdf_converter.rb [hunk 2]
```diff
@@ -1,31 +1,58 @@
...
-        raise FileConversionError
-      end
+    attachment.open do |file|
+      next if DocxFile.has_embedded_files?(file)
+
+      command = "soffice --headless --convert-to pdf #{file.path} --outdir #{dir}"
+      raise FileConversionError unless Kernel.system(command)
+
+      filename_without_ext = File.basename(file.path, ".*")
+      output_pdf_path = "#{dir}/#{filename_without_ext}.pdf"
```
[the remark this answers is not in the document]
Comment (FerPerales) [reply app/services/doc_to_pdf_converter.rb:31]: @jutonz we are using Heroku, which uses an ephemeral file system that, if I recall correctly, get deletes everytime the dyno restarts. We can confirm this behavior in staging and see if we need to remove the files manually

File: spec/factories/source_files.rb [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
       standards_import = create(:standards_import, :with_files)
       SourceFile.where(
         active_storage_attachment: standards_import.files.first
       ).first
     end
 
-    trait :with_docx_attachment do
+    traits_for_enum :status, SourceFile.statuses
+
+    trait :docx do
```
[the remark this answers is not in the document]
Comment (FerPerales) [reply spec/factories/source_files.rb:12]: @jutonz this makes sense! Thanks!

**Review by FerPerales — APPROVED** [review]
@jutonz looks good! Let's have @littleforest giving another review and we are good to go 🚀

**Review by littleforest — COMMENTED** [review]
@jutonz thanks for working on this! Code looks good, but my only concern is that the PDF file will automatically get placed as the redacted file. @FerPerales will the admin operators still be able to further redact that pdf file? What if they mess up and want to remove the redacted file? I think we had discussed that we like having the PDF icon for the "View original document" and wanted to keep that consistent with PDFs throughout.

Screenshot 2024-01-24 at 9 57 51 AM

**Comment by FerPerales** [comment]
> @jutonz thanks for working on this! Code looks good, but my only concern is that the PDF file will automatically get placed as the redacted file. @FerPerales will the admin operators still be able to further redact that pdf file? What if they mess up and want to remove the redacted file? I think we had discussed that we like having the PDF icon for the "View original document" and wanted to keep that consistent with PDFs throughout.

They won't be able to redact the file, but they can redact the docx file prior to this process, In case this process already created a PDF as the redacted file, we now have a feature that allow admins to remove the redacted file. We can check with Patrick what would be the expected flow and edge cases

**Comment by littleforest** [comment]
> They won't be able to redact the file, but they can redact the docx file prior to this process

I don't think the operators are set up to do redaction except through the Foxit redaction process with the PDF. I'm pretty sure we want a consistent process in place. Is it possible to create a new file attachment on the SourceFile that can link to the Word doc, and then we can move the converted PDF to the main attachment?

>  In case this process already created a PDF as the redacted file, we now have a feature that allow admins to remove the redacted file

If they remove this [redacted] PDF file, then will there be a process to get the original Word Doc re-converted to PDF?

**Comment by jutonz** [comment]
Ok cool, I can start working on the system you described @littleforest. I don't think that's a massive change, and there are other things I can reference that work similarly. Just to confirm, here's what I'm thinking:

1. Find source files that have a docx attachment without any embedded files
2. Create a new SourceFile and attach the original docx file 
    * In the [future](https://app.asana.com/0/1203289004376659/1206228142308910) this SourceFile will be marked as archived
3. Convert the docx file into a pdf, and update the attachment on the original SourceFile to point to that pdf
    * Setup like an `original_source_file` association on the new SourceFile to point at the original docx? 

Does that sound reasonable? Also happy to hop on a call if that's easier to chat about this!

File: app/services/doc_to_pdf_converter.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jutonz) [reply app/services/doc_to_pdf_converter.rb:31]: ok cool that makes sense and is good to hear! thanks!

**Comment by jutonz** [comment]
Ok cool I think this should be ready for another round of review! Instead of using `redacted_file` for the converted PDF we now create a totally separate SourceFile. This is done indirectly by appending the new PDF to `standards_import.files`, which has the effect of creating a new SourceFile. The new SourceFile will have its `original_source_file` association set to the original docx SourceFile, which will itself be marked as archived. 

Linking the new and original source files was a little awkward because SourceFile creation happens as a callback on StandardsImport, so we don't have the ability to pass args or anything into that process to indicate that we want two source files linked. Instead this sets a `link_to_pdf_filename` field on the original SourceFile, which the `CreateSourceFilesJob` will use to link up the new SourceFile once it creates it. 

A future refactor which might make this a bit simpler to do would be to pull the `CreateSourceFilesJob` out of a callback on StandardsImport and instead do that logic manually via a service object or something like that. That would allow us to have a bit more control over where and how SourceFiles get created. I didn't do that here since this PR was already getting big, but maybe something to consider for the future :)

**Review by littleforest — APPROVED** [review]
Looks great! Thanks for making those changes.

**Review by FerPerales — APPROVED** [review]
:shipit:

**Pull request merged**